server/src/main/java/io/crate/expression/reference/sys/check/cluster/NumberOfPartitionsSysCheck.java

```diff
@@ -60,7 +64,7 @@ public boolean isValid() {
     boolean validateDocTablesPartitioning(SchemaInfo schemaInfo) {
         for (TableInfo tableInfo : schemaInfo.getTables()) {
             DocTableInfo docTableInfo = (DocTableInfo) tableInfo;
-            if (docTableInfo.isPartitioned() && docTableInfo.partitions().size() > PARTITIONS_THRESHOLD) {
+            if (docTableInfo.isPartitioned() && docTableInfo.getPartitions(clusterService.state().metadata()).size() > PARTITIONS_THRESHOLD) {
```

Comment:
Could lift the `clusterState.state()` call out of the loop to avoid repeated access on the volatile variable.

I'd also inline the second function - to make the `DocSchemaInfo` -> `DocTableInfo` relationship a bit clearer. Otherwise one doesn't immediate get why the cast is safe.

```java
    @Override
    public boolean isValid() {
        Metadata metadata = clusterService.state().metadata();
        for (SchemaInfo schemaInfo : schemas) {
            if (!(schemaInfo instanceof DocSchemaInfo docSchemaInfo)) {
                continue;
            }
            for (var table : docSchemaInfo.getTables()) {
                DocTableInfo docTableInfo = (DocTableInfo) table;
                if (docTableInfo.isPartitioned() && docTableInfo.getPartitions(metadata).size() > PARTITIONS_THRESHOLD) {
                    return false;
                }
            }
        }
        return true;
    }
```

Reply:
Thank you, added an extra commit for this.